aopy/data/bmi3d.py

```diff
@@ -832,6 +835,46 @@ def get_target_locations(preproc_dir, subject, te_id, date, target_indices):
             raise ValueError(f"Target index {target_indices[i]} not found")
     return np.round(locations,4)
 
+def get_trajectory_frequencies(preproc_dir, subject, te_id, date):
+    # load task codes
+    config_dir = files('aopy').joinpath('config')
+    params_file = as_file(config_dir.joinpath('task_codes.yaml'))
+    with params_file as f:
+        task_codes = yaml_read(f)[0]
+
+    # load preproc data
+    data, metadata = load_preproc_exp_data(preproc_dir, subject, te_id, date)
+    params = json.loads(metadata['sequence_params'])
+    primes = np.asarray(list(sympy.primerange(0, sympy.prime(params['num_primes'])+1)))
+    even_idx = np.arange(len(primes))[0::2]
+    odd_idx = np.arange(len(primes))[1::2]
+    base_period = 20
+
+    # recreate random trial order of reference & disturbance frequencies
+    np.random.seed(params['seed'])
+    order = np.random.choice([0,1])
+    if order == 0:
+        trial_r_idx = np.array([even_idx, odd_idx]*params['ntrials'], dtype='object')
+        trial_d_idx = np.array([odd_idx, even_idx]*params['ntrials'], dtype='object')
+    elif order == 1:
+        trial_r_idx = np.array([odd_idx, even_idx]*params['ntrials'], dtype='object')
+        trial_d_idx = np.array([even_idx, odd_idx]*params['ntrials'], dtype='object')
+        
+    # get trajectory generator index used for each trial
+    trial_start_codes = [task_codes['CENTER_TARGET_ON']]
+    trial_end_codes = [task_codes['REWARD'], task_codes['OTHER_PENALTY'], 
+                       task_codes['HOLD_PENALTY'], task_codes['TIMEOUT_PENALTY']] # using TRIAL_END may not pass assertion
+    gen_trajectories, _ = get_kinematic_segments(preproc_dir, subject, te_id, date, 
+                                                 trial_start_codes, trial_end_codes, datatype='gen_idx', 
+                                                 samplerate=metadata['fps'], filter_kinematics=False)
+    assert np.all([int(gen[0])==int(gen[-1]) for gen in gen_trajectories])
+    gen_idx = [int(gen[0]) for gen in gen_trajectories]
+
+    # use generator index to get reference & disturbance frequencies for each trial
+    freq_r = [primes[np.array(idx, dtype=int)]/base_period for idx in trial_r_idx[gen_idx]]
+    freq_d = [primes[np.array(idx, dtype=int)]/base_period for idx in trial_d_idx[gen_idx]]
+    return freq_r, freq_d
```

Comment:
could all of this be moved into a function inside `../preproc/bmi3d.py`? This is no longer a `data` function

Reply:
Does it make sense to just move the entire `get_trajectory_frequencies()` function to preproc/bmi3d.py?

Reply:
not the entire thing because it loads the file from preproc/subject/etc. the preproc functions generally load exp_data, exp_metadata, etc.